feat(hardware/gyro): implement gyroscopes

Baconing opened the pull request:
Implements IMUs and Single Axis gyroscopes.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced the `ADIS16448IMU`, `ADIS16470IMU`, and `NavXIMU` classes for enhanced inertial measurement capabilities.
	- Integrated the `wpiunits-java` library to improve unit handling in robotics programming.
	- Added comprehensive unit tests for `ADIS16448IMU`, `ADIS16470IMU`, `IMU`, and `NavXIMU` classes to ensure functionality and reliability.

- **Bug Fixes**
	- Removed the outdated `SingleAxisGyroscope` class, streamlining the codebase.

These updates enhance the functionality and performance of the robotics applications by providing more robust IMU options, improved unit management, and thorough testing coverage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Baconing committed:
refactor(hardware/gyro): separate IMUs and single axis gyros

Baconing committed:
feat(hardware/gyro/imu): implement ADIS16448IMU and ADIS16470IMU

Baconing committed:
refactor(hardware/gyro/imu): remove NotNull annotations from abstract methods

Baconing committed:
refactor(hardware/gyro/imu): use Translation3d instead of Rotation3d for acceleration.

Baconing committed:
refactor(hardware/gyro/imu): use radians instead of degrees

Baconing committed:
feat/refactor(hardware/gyro/imu/adis16*): finish implementations

Baconing committed:
style(hardware/gyro/imu/adis16470): remove unused import

Baconing committed:
chore(build.gradle): add wpiunits dependency

Baconing committed:
refactor(hardware/gyro/imu/adis16470): use unqualified name instead of FQDN in switch statement

Baconing commented on src/main/java/net/frc5183/librobot/hardware/gyro/imu/ADIS16470IMU.java:
```diff
@@ -0,0 +1,148 @@
+    /**
+     * The YAW axis.
+     */
+    private final IMUAxis yaw;
+
+    /**
+     * The PITCH axis.
+     */
+    private final IMUAxis pitch;
+
+    /**
+     * The ROLL axis.
+     */
+    private final IMUAxis roll;
```
@Trip-kun thoughts?

Trip-kun replied:
This solution certainly isn't perfect. Auto adjusting the axis to account for physical rotation in the bot is good. I don't know if this is complete though. Optimally would be applying a rotation 3d to the outputs to correct for any possible rotation but nobody wants to calculate a rotation 3d by hand. I think as long as try and support all rotations it should be okay. (by a rough think there should be 24 90 degree based rotations.

Baconing replied:
Are you talking about offsets? I don't really understand

Baconing replied:
Anyway, I realized that the constructor for Rotation3d is actually asking for yaw pitch roll not XYZ.

Baconing commented on src/main/java/net/frc5183/librobot/hardware/gyro/imu/ADIS16448IMU.java:
```diff
@@ -0,0 +1,168 @@
+    public ADIS16448IMU(@NotNull ADIS16448_IMU imu) {
+        yaw = toIMUAxis(imu.getYawAxis());
+
+        //todo: confirm this is correct i had copilot do it
+        if (yaw == IMUAxis.X) {
+            pitch = IMUAxis.Y;
+            roll = IMUAxis.Z;
+        } else if (yaw == IMUAxis.Y) {
+            pitch = IMUAxis.X;
+            roll = IMUAxis.Z;
+        } else {
+            pitch = IMUAxis.X;
+            roll = IMUAxis.Y;
+        }
```
@Trip-kun does this look good to you?

Baconing commented on src/main/java/net/frc5183/librobot/hardware/gyro/imu/ADIS16448IMU.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    @Override
+    public @NotNull Translation3d getAccelerationMetersPerSecondSquared() {
+        return new Translation3d(imu.getAccelX(), imu.getAccelY(), imu.getAccelZ());
+    }
```
These functions return in m/s^2

Baconing committed:
refactor(hardware/gyro/imu): put unit in method name for IMU.getRotation3d() and IMU.getRawRotation3d()

Trip-kun commented on src/main/java/net/frc5183/librobot/hardware/gyro/imu/IMU.java:
```diff
@@ -0,0 +1,77 @@
+     */
+    public abstract void reset();
+
+    /**
+     * Returns the raw gyroscope object.
+     * Should only be used when you know the subclass of the gyroscope.
+     * <br><br>
+     * Implementations should override the return type to the type of the gyroscope used in the vendor's API.
+     * (e.g the {@link ADIS16448IMU} would return {@link edu.wpi.first.wpilibj.ADIS16448_IMU})
+     * @return the raw gyroscope object.
+     */
+    public abstract Object getRawIMU();
```
This is not a great solution because it means everything that uses a gyroscope has to either no longer be blind yk the type of the gyroscope or has to also be a generic.
The getRawIMU method is fine for what it is. Just getting the object and type checking it to be the correct wpilib imu type

Baconing replied:
#10

Baconing commented on src/main/java/net/frc5183/librobot/hardware/gyro/imu/ADIS16470IMU.java:
```diff
@@ -0,0 +1,148 @@
+        imu.calibrate();
+    }
+
+    @Override
+    public void reset() {
+        imu.reset();
+    }
+
+    @Override
+    public @NotNull ADIS16470_IMU getRawIMU() {
+        return imu;
+    }
+
+    /**
+     * Returns a new {@link ADIS16470_IMU.IMUAxis} from an {@link IMUAxis}.
+     * @param axis the {@link IMUAxis} to convert.
+     * @return the converted {@link ADIS16470_IMU.IMUAxis}.
+     */
+    public static ADIS16470_IMU.IMUAxis fromIMUAxis(IMUAxis axis) {
+        return switch (axis) {
+            case X -> ADIS16470_IMU.IMUAxis.kX;
+            case Y -> ADIS16470_IMU.IMUAxis.kY;
+            case Z -> ADIS16470_IMU.IMUAxis.kZ;
+        };
+    }
+
+    /**
+     * Returns a new {@link IMUAxis} from an {@link ADIS16470_IMU.IMUAxis}.
+     * @param axis the {@link IMUAxis} to convert.
+     * @return the converted {@link ADIS16470_IMU.IMUAxis}.
+     */
+    public static IMUAxis toIMUAxis(ADIS16470_IMU.IMUAxis axis) {
+        return switch (axis) {
+            case kX -> IMUAxis.X;
+            case kY -> IMUAxis.Y;
+            case kZ -> IMUAxis.Z;
+            default -> throw new IllegalArgumentException("IMU axis must be one of  " + axis);
+        };
+    }
+}
```
They'll serve purpose in later commits when SingleAxisGyroscope is implemented.

Baconing committed:
feat(hardware/gyro/imu): add getAngleRadians(SingleAxisGyroscope.Axis) method to IMU

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
refactor(hardware/imu/adis16448): use unqualified name instead of FQDN in switch statement

Baconing committed:
Merge remote-tracking branch 'origin/v2025-1-0-0/gyros' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Baconing committed:
feat/fix(hardware/gyro/imu/adis16*): implement getAngleRadians(SingleAxisGyroscope.Axis) and fix getRawRotation3dRadians() by using yaw pitch roll instead of XYZ

Baconing committed:
feat(hardware/gyro/imu): finish implementation of IMU and it's subclasses

Baconing committed:
refactor(hardware/gyro/imu): move yaw pitch roll axis assignment to a utility method

Baconing commented on src/main/java/net/frc5183/librobot/hardware/gyro/imu/ADIS16448IMU.java:
```diff
@@ -0,0 +1,258 @@
+        super.setOffset(new Rotation3d());
+    }
+
+    @Override
+    public void clearStickyFaults() {
+        // The ADIS16448 IMU does not have a method for clearing sticky faults.
+    }
```
The function should not do anything under this circumstance, as there are no side effects from this function not doing anything on the ADIS16448.

Baconing committed:
refactor(hardware/gyro): remove SingleAxisGyroscope, just use IMU.

all the functionality from SingleAxisGyroscope exists in IMU, and it just complicates things

Baconing committed:
feat(hardware/gyro): implement KauaiLabs NavX

Baconing committed:
docs(hardware/gyro/adis16*): update documentation to reflect changes

Changes reflected:
IMUAxis -> CartesianAxis
SingleAxisGyroscope -> Attitude

Baconing committed:
style(hardware/gyro/imu): remove unnecessary qualifiers in IMU.CartesianAxis.assignAxes(CartesianAxis yaw)

Baconing committed:
refactor(hardware/gyro): clean up some of the long methods

Trip-kun commented on src/main/java/net/frc5183/librobot/hardware/gyro/IMU.java:
I would run back over the getAngleRadians method. It may be better to "apply" the rotation to the entire 3d rotation and then extract the axis you want rather than just adding the offset to the raw output. I'm not entirely sure which would be correct here but it is notable that 3d chained rotations are complicated (there's a reason they can be done with quaternions and quaternion multiplication is not communicative.

Trip-kun replied:
https://math.stackexchange.com/questions/2016937/why-is-the-product-of-two-rotation-matrices-not-commutative

an example of what I'm talking about, by the way

Trip-kun commented on src/main/java/net/frc5183/librobot/hardware/gyro/SingleAxisGyroscope.java:
I would not just remove single axis gyroscopes. They can be useful for tracking rotating parts like arms or grabbers. In addition, you could "get" a single axis gyroscope from IMUs to allow more adaptability.

Baconing committed:
Merge branch 'master' into v2025-1-0-0/gyros

Signed-off-by: Brenden Freier <iam@baconi.ng>

Baconing committed:
refactor(hardware/gyro/imu): only use Rotation3d where absolutely necessary

keep my family's name out of your mouth, quaternions

Baconing committed:
test(hardware/gyro): implement tests for ADIS16448, ADIS16470, and default IMU methods.

ADIS16470 acceleration tests are broken due to a [bug in wpilibj](https://github.com/wpilibsuite/allwpilib/issues/7180).

Baconing committed:
fix(hardware/gyro): implement getRate() function

Baconing committed:
fix(hardware/gyro): (re)implement angle rate logic

Baconing committed:
test(hardware/gyro): finish implementing unit tests for all implemented gyroscopes

Baconing committed:
fix(hardware/gyro/navx): use Gs not m/ss

Baconing commented on src/main/java/net/frc5183/librobot/hardware/gyro/ADIS16448IMU.java:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    @Override
+    public @NotNull Translation3d getAccelerationMetersPerSecondSquared() {
+        return new Translation3d(imu.getAccelX(), imu.getAccelY(), imu.getAccelZ());
+    }
```
ADIS16448 and 16470 use m/ss for acceleration

Baconing committed:
docs(hardware/gyro): annotate deprecated members

Baconing commented on src/test/java/net/frc5183/librobot/hardware/gyro/NavXTests.java:
```diff
@@ -0,0 +1,639 @@
+         * NavX simulation does not update on-demand, so we have to wait until it updates next.
+         * I could grab this value programmatically, but I'm lazy and the current delay seems fine.
+         */
+        private void delay() {
+            try {
+                Thread.sleep(waitDuration);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
```
What methods would you recommend other than a while loop? Most won't work due to the loss of precision between the AHRS and simulation communication

Baconing replied:
Well, to start, you have to get the value from the actual AHRS object, not the simulation value, as we're waiting for updates on the AHRS. Secondly, if you add the AHRS as an instance variable and get its value, you also need to apply a delta to account for precision loss in communication between the simulator and AHRS object. You also need to redefine this function for each variable being updated.

Baconing commented on src/test/java/net/frc5183/librobot/hardware/gyro/NavXTests.java:
```diff
@@ -0,0 +1,639 @@
+    static final double TEST_ANGLE_2 = 90;
+    static final double TEST_ANGLE_3 = 180;
+    static final double TEST_ANGLE_4 = 270;
+    static final double TEST_ANGLE_5 = Math.random() * 360;
```
This is to ensure that values aren't just hard coded, and theoretically any value should produce the same result for these tests.

Baconing commented on src/test/java/net/frc5183/librobot/hardware/gyro/ADIS16448Tests.java:
```diff
@@ -0,0 +1,639 @@
+        sim.setAccelZ(TEST_ACCEL_4);
+        assertEquals(TEST_ACCEL_4, imu.getAccelerationMetersPerSecondSquared(IMU.CartesianAxis.Z));
+
+        sim.setAccelZ(TEST_ACCEL_5);
+        assertEquals(TEST_ACCEL_5, imu.getAccelerationMetersPerSecondSquared(IMU.CartesianAxis.Z));
+    }
+
+    @Test
+    void testAccelerationTranslation3d() {
+        // Meters per second squared.
+        Consumer<Double> setAccel = (Double accel) -> {
+            sim.setAccelX(accel);
+            sim.setAccelY(accel);
+            sim.setAccelZ(accel);
+        };
+
+        // Meters per second squared.
+        BiConsumer<Double, Double> testAccel = (Double accel, Double expected) -> {
+            assertEquals(
+                    new Translation3d(expected, expected, expected),
+                    imu.getAccelerationMetersPerSecondSquared(),
+                    "Translation3d failed."
+            );
+        };
+
+        setAccel.accept(TEST_ACCEL_1);
+        testAccel.accept(TEST_ACCEL_1, TEST_ACCEL_1);
+
+        setAccel.accept(TEST_ACCEL_2);
+        testAccel.accept(TEST_ACCEL_2, TEST_ACCEL_2);
+
+        setAccel.accept(TEST_ACCEL_3);
+        testAccel.accept(TEST_ACCEL_3, TEST_ACCEL_3);
+
+        setAccel.accept(TEST_ACCEL_4);
+        testAccel.accept(TEST_ACCEL_4, TEST_ACCEL_4);
+
+        setAccel.accept(TEST_ACCEL_5);
+        testAccel.accept(TEST_ACCEL_5, TEST_ACCEL_5);
+    }
```
How can we access the sim and imu instance variables in the static function?

Baconing replied:
This fails with error `org.junit.platform.commons.PreconditionViolationException: Cannot invoke non-static method [java.util.stream.Stream<org.junit.jupiter.params.provider.Arguments> net.frc5183.librobot.hardware.gyro.NavXTests.provideCartesianAccelerationsForTesting()] on a null target.`
